legend-engine-authentication-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/AuthenticationGrammarParserExtension.java

```diff
@@ -0,0 +1,130 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.finos.legend.engine.language.pure.grammar.from;
+
+import org.antlr.v4.runtime.CharStream;
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.finos.legend.engine.language.pure.grammar.from.extensions.IAuthenticationGrammarParserExtension;
+import org.finos.legend.engine.language.pure.grammar.from.authentication.AuthenticationParseTreeWalker;
+import org.finos.legend.engine.language.pure.grammar.from.antlr4.authentication.AuthenticationLexerGrammar;
+import org.finos.legend.engine.language.pure.grammar.from.antlr4.authentication.AuthenticationParserGrammar;
+import org.finos.legend.engine.language.pure.grammar.from.antlr4.authentication.CredentialLexerGrammar;
+import org.finos.legend.engine.language.pure.grammar.from.antlr4.authentication.CredentialParserGrammar;
+import org.finos.legend.engine.language.pure.grammar.from.authentication.AuthenticationSourceCode;
+import org.finos.legend.engine.protocol.pure.v1.model.packageableElement.authentication.AuthenticationSpec;
+import org.finos.legend.engine.protocol.pure.v1.model.packageableElement.authentication.Credential;
+import org.finos.legend.engine.language.pure.grammar.from.authentication.CredentialParseTreeWalker;
+import org.finos.legend.engine.language.pure.grammar.from.authentication.CredentialSourceCode;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+
+public class AuthenticationGrammarParserExtension implements IAuthenticationGrammarParserExtension
+{
+    public static final String NAME = "Authentication";
+
+    public List<Function<AuthenticationSourceCode, AuthenticationSpec>> getExtraAuthenticationParsers()
+    {
+        return Collections.singletonList(code ->
+        {
+            SourceCodeParserInfo parserInfo = getAuthenticationParserInfo(code);
+            AuthenticationParseTreeWalker walker = new AuthenticationParseTreeWalker(parserInfo.walkerSourceInformation);
+            switch (code.getType())
+            {
+                case "ApiKeyAuthenticationSpec":
+                    return parseAuthenticationSpec(code, p -> walker.visitApiKeyAuthenticationSpec(code,p.apiKeyAuthentication()));
+                case "UsernamePasswordAuthenticationSpec":
+                    return parseAuthenticationSpec(code, p -> walker.visitUsernamePasswordAuthenticationSpec(code, p.basicAuthentication()));
+                case "OauthAuthenticationSpec":
```

Comment:
for grammar we should try to keep those short, so maybe we can exclude AuthenticationSpec

Reply:
Done